sw/host/opentitanlib/src/transport/verilator/uart.rs

```diff
@@ -154,4 +155,18 @@ impl Uart for VerilatorUart {
         self.file.borrow_mut().seek(SeekFrom::End(0))?;
         Ok(())
     }
+
+    fn as_console_device(&self) -> &dyn ConsoleDevice {
+        self
+    }
+}
+
+impl ConsoleDevice for VerilatorUart {
```

Comment:
Shouldn't any implementation of the `Uart` trait also be an implementation of the `ConsoleDevice`?  I dislike the need for each implementation to also need to have these forwarding functions.

I do not know enough Rust to be able to tell if traits can "inherit" from each other, if they can, then both the `Uart` trait and `spi::Target` trait should be modified to "inherit from / extend" the `ConsoleDevice` trait, and then you would not need these.

Alternatively, leave the `Uart` and `spi::Target` traits as is, and do not change `VerilatorUart` or other implementations of these traits to implement `ConsoleDevice`, but instead create two new structs: `UartConsoleDevice::new(Rc<&dyn Uart>)`, and `SpiConsoleDevice::new(Rc<&dyn spi::Target>)`. That way, these forwarding functions will only have to be implemented once, and not for each implementation of either `spi::Target` or `Uart`.

Reply:
One reason I am suggesting the above is that I foresee some merge conflict with my PR #17271, introducing non-blocking Uart reads.  In order to adapt my PR, I would probably need to add a few more non-blocking related methods to the `ConsoleDevice` trait, and it is undesirable to at that point have to touch every interface implementation again, when all that I want to express is how one trait (`ConsoleDevice`) builds on top of another trait (`Uart`).

Reply:
I am using a Rust `supertrait` to define the `Uart` and SPI `Target` traits, hence in this PR you see `Uart: ConsoleDevice`. However according to the [Rust supertrait docs](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-supertraits-to-require-one-traits-functionality-within-another-trait), you still must `impl` both the trait and its supertrait on the target structs. According to the [docs](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch17-01-what-is-oo.html#inheritance-as-a-type-system-and-as-code-sharing): _"There is no way to define a struct that inherits the parent struct’s fields and method implementations."_

Reply:
The reason I decided to make a ConsoleDevice trait that must be implemented for each UART/SPI is because depending on the transport the implementations may be different. It just so happens that for UART all the transports have the same implementation, however for other potential console devices, e.g. SPI, things might be different. Namely, not every transport may have a SPI, so the default "not supported" implementations will suffice, or in the case of the HyperDebug SPI console `console_read` implementation, we have to do a write followed by a read to get around the hyp firmware issues we discussed last week (see #17817 for the implementation details).

Reply:
Thanks for explaining the super-traits.  It seems that they do not solve the problem the way that I had hoped.

My suggestion of having `SpiConsoleDevice::new(Rc<&dyn spi::Target>)` would also work nicely for transports that do not support SPI at all, the caller would not be able to get a `spi::Target` instance.  As for differences in SPI protocol between implementations:  I see that in PR #17817 you have implemented a little protocol that allows querying how many bytes of data is available (I would have preferred that the first four bytes could be used to distinguish between retrieving buffer status, and the actual contents, with a third future value for sending data, though I have not looked into how to make it optimally match the TPM device capabilities).  My point is that even if I wanted to use Dediprog, UltraDebug, or some other transport to access the OpenTitan SPI console, I would have to use the same protocol, so in my view, that protocol should go into `SpiConsoleDevice`, and be shared between transports, rather than repeated in their implementation.

Reply:
> One reason I am suggesting the above is that I foresee some merge conflict with my PR #17271, introducing non-blocking Uart reads. In order to adapt my PR, I would probably need to add a few more non-blocking related methods to the `ConsoleDevice` trait, and it is undesirable to at that point have to touch every interface implementation again, when all that I want to express is how one trait (`ConsoleDevice`) builds on top of another trait (`Uart`).

Ah got it, thanks for explaining. Let me give your suggestion a try.

Reply:
@jesultra I refactored this further, lmk what you think now.

Reply:
SpiConsoleDevice struct is added in #17817.

Reply:
Also, framing protocol (generated on the device side) is implemented in #17818.